### ISSUE TITLE
Clique PoA implementation plan

### ISSUE THREAD

**Issue by gumb0** [ethereum/aleth#5522]
## Block validation

These checks should go into `verify` method of Clique's implementation of `SealEngineFace`.
- mixhash is 0
- uncles list is empty
- timestamp >= parent.timestamp + 15s
- difficulty is either 1 or 2, depending on signer index
- check that signer's last signed block is more than SIGNER_LIMIT blocks ago
- For epoch transition block
	- beneficiary is 0
	- nonce is 0
	- check the list of signers in the block header is equivalent to our list
- For regular blocks
	- nonce is either 0 or 0xff..f
	- extra-data size is 32 + 65 bytes

## Keeping track of signers

We should have a list of currently authorized signers and the list of current pending votes. The block processing has an additional step of processing votes.

- Initializing signer list from extra-data of genesis block
- Processing of regular block
	- Update vote list
	- if proposal reaches SIGNER_LIMIT, modify signers list, and discard all votes for this beneficiary
- Processing of the epoch transition block
	- Discard all pending votes
- Serializing signers and votes
- Tests for signer list logic listed in EIP

## CLI

- Config for Goerli
- Config for Rinkeby
- Command line arguments --goerli and --rinkeby

## (optional) Support Goerli in Warp sync

**Comment by halfalicious** [comment]
Oh awesome, I was just thinking of putting together a plan for this. Great timing!

**Comment by axic** [comment]
cc @jwasinger

**Comment by jwasinger** [comment]
A note on sealing (and block import) implementation:

The block coinbase field contains the vote (or is zeroed).  During contract execution, it must be set to the author of the block and reset to the original value (the vote) after contract execution (and before the block is signed).  Difficulty must also be set before contract execution.

**Comment by gumb0** [comment]
@jwasinger yeah I've mentioned coinbase change in "Other important things" part.
And `DIFFICULTY` opcode I think will work the same as for Ethash (will get it from the header)

**Comment by jwasinger** [comment]
My point was mainly to illustrate that you need to set the `COINBASE` to a different value when executing contracts, and reset it afterwards to the vote value.

**Comment by axic** [comment]
@jwasinger if this is not clarified in https://eips.ethereum.org/EIPS/eip-225 I suggest to create a PR improving the EIP.

**Comment by jwasinger** [comment]
Good point.  Done https://github.com/ethereum/EIPs/pull/1846

**Comment by gumb0** [comment]
> Serializing signers and votes - the lists should be saved into DB periodically (the simplest solution - save them for each block, though this might be wasteful)

I've read through Parity's implementation and they do something simpler, though probably less efficient - just having an in-memory cache of vote states for a number (128) of blocks. In case required state is not present in the cache, it is reconstructed from the latest checkpoint block (by replaying all votes in blocks between checkpoint and required one)

I think it would be easier for us to start with something similar, too.

**Comment by gumb0** [comment]
Updated `Keeping track of signers` section now with a better understanding.

**Comment by gumb0** [comment]
Added a bit of more info in other sections, too.

**Comment by gumb0** [comment]
Some highlights of Parity's implementation that I find important/relevant for us:

PR with all changes https://github.com/paritytech/parity-ethereum/pull/9981

Higher-level description of how the code works
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/mod.rs#L17-L59

### Keeping track of signers

Data structure to keep the Clique signers/voting state
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/block_state.rs#L60

Processing the votes or checkpoint block (update Clique state from the new header)
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/block_state.rs#L163

Cache storing 128 recently used Clique states
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/mod.rs#L164

Function that reconstructs Clique state from checkpoint block if not found in cache
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/mod.rs#L272

### Block validation
Validating the header
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/mod.rs#L536

### Sealing
Filling the block to seal
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/mod.rs#L381
Another piece of filling the block, including setting the difficulty field
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/mod.rs#L699-L735

Calculating the timestamp when to seal the next block (including random delay for out-of-turn seal)
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/block_state.rs#L268

### Other
Optimization for recovery of signer addresses from block signatures: cache that stores recently recovered signer addresses
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/util.rs#L34

Signer address recovery function
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/util.rs#L38

Function to extract the signer list from checkpoint block header
https://github.com/paritytech/parity-ethereum/blob/78d0a8696ffec4eb6b0e2214b476fed388e14bed/ethcore/src/engines/clique/util.rs#L85
